src/browser/ui/dom/components/__tests__/ReactDOMInput-test.js

```diff
@@ -299,30 +303,41 @@ describe('ReactDOMInput', function() {
       React.render(<input type="checkbox" checkedLink={link} />, node);
       expect(console.warn.mock.calls.length).toBe(0);
 
-      React.render(
+      ReactTestUtils.renderIntoDocument(
         <input
           type="checkbox"
           checked="false"
           onChange={mocks.getMockFunction()}
-        />,
-        node
+        />
       );
       expect(console.warn.mock.calls.length).toBe(0);
 
-      React.render(
-        <input type="checkbox" checked="false" readOnly={true} />,
-        node
+      ReactTestUtils.renderIntoDocument(
+        <input type="checkbox" checked="false" readOnly={true} />
       );
       expect(console.warn.mock.calls.length).toBe(0);
 
-      React.render(<input type="checkbox" checked="false" />, node);
+      ReactTestUtils.renderIntoDocument(<input type="checkbox" checked="false" />);
       expect(console.warn.mock.calls.length).toBe(1);
+    } finally {
+      console.warn = oldWarn;
+    }
+  });
+
+  it('should warn with checked and no onChange handler with readOnly specified', function() {
+    var oldWarn = console.warn;
+    try {
+      console.warn = mocks.getMockFunction();
+
+      ReactTestUtils.renderIntoDocument(
+        <input type="checkbox" checked="false" readOnly={true} />
+      );
+      expect(console.warn.mock.calls.length).toBe(0);
 
-      React.render(
-        <input type="checkbox" checked="false" readOnly={false} />,
-        node
+      ReactTestUtils.renderIntoDocument(
+        <input type="checkbox" checked="false" readOnly={false} />
       );
-      expect(console.warn.mock.calls.length).toBe(2);
+      expect(console.warn.mock.calls.length).toBe(1);
```

Comment:
There is a minor functional change, in that we will only fire each warning once (otherwise, you'd get the same warning over and over again every time a render() was called, which would be very noisy).  This means that in these two places where we were expecting console.warn.mock.calls.length to be 2, it will only ever reach 1.  I refactored the code into a separate test so that we can assert the warning does fire on the condition being tested.  See lines 266-269 of ReactElementValidator.js